Problem with DerivativeForm Constructor

I am trying to understand the proper instantiation of a `DerivativeForm<2, dim, spacedim, Number>`; as far as I understood it can be interpreted as a rank 3 tensor describing the hessian **HF** of a mapping **F**. And from the documentation it is mentioned that it can be thought of a matrix with dimensions `(spacedim x dim x dim)`. 

Then, when using `DerivativeForm<2, dim spacedim, Number>::operator[](const unsigned int)` an object of type `Tensor<2, dim, Number>` is returned. In other words, we get a `dim x dim` matrix written as a rank 2 tensor. 

However, for the constructor we hand an object of type (suppose `dim != spacedim`) `Tensor<2, spacedim, Tensor<1,  dim> >` to build **HF** out of. Using `Tensor<2, spacedim, Tensor<1, dim>>::operator[](const unsigned int)`, an object of type `Tensor<1, spacedim, Tensor<1, dim>>` will be returned for that object, that represents a rank 2 tensor with dimensions `spacedim x dim`. 

In other words, the object given to the constructor of `DerivativeForm<2, dim, spacedim, Number>` of type `Tensor<2, spacedim, Tensor<1,  dim> >` seems to describe a rank 3 tensor similar to a `(spacedim x spacedim x dim)`-matrix. This is also correct from the documentation:

>  A rank-2 tensor is a linear operator that can transform a vector into another vector. These tensors are similar to matrices with dim×dim components.

This leads me to believe, that there is a dimension mismatch for `DerivativeForm<order, dim, spacedim>`, whenever `dim != spacedim`, and I suppose the correct object to hand to the constructor would be `Tensor<1, spacedim, Tensor<order,  dim> >` at least for `order = 2`. 

With the above, I still try to comprehend why the compiler returns no errors at this point: 
```
template <int order, int dim, int spacedim, typename Number>
 inline DerivativeForm<order, dim, spacedim, Number>::DerivativeForm(
   const Tensor<order, spacedim, Tensor<1, dim, Number>> &T)
 {
   for (unsigned int j = 0; j < spacedim; ++j)
     (*this)[j] = T[j];
 }
```
In my opinion the data types on the left- and right-hand-side are incompatible, but it compiles just fine. 

Is there anyone, who might shed some light on this topic? 

Edit: The more I think about it, the desired input just makes no sense: Mathematically, we can think of the Hessian of a function F : \R^dim \to \R^spacedim as a rank 3 tensor **HF** with dimensions `spacedim x dim x dim` as mentioned in the documentation, where we collect for each F_i the hessian **HF_i**. The data type seems to relate to a `Tensor<1, spacedim, Tensor<2, dim> >`. On the other hand, we can also identify it with a `dim x dim` matrix, where each entry is a `spacedim` vector, i.e. **HF** can be viewed as a `dim x dim x spacedim` matrix, where we can use the relating data type `Tensor<2, dim, Tensor<1, spacedim> >`. In this case the sizes `dim` and `spacedim` are switched compared to the needed type `Tensor<2, spacedim, Tensor<1, dim> >`. 
<!--
Thanks for reporting an issue with deal.II! This HTML comment won't be visible
in the issue tracker but is present here to help you write a better description
of the problem you encountered.

If you have questions about deal.II, please post them on the mailing list:

   https://groups.google.com/g/dealii

Otherwise:
- Do you encounter a compilation failure? Please attach the detailed.log file
  present in the build directory here. Any additional compiler messages will also
  be helpful!

- Does something crash (e.g., an example) that shouldn't? Please include any relevant
  terminal output as well as the detailed.log file generated in your build
  directory (should you have one).

- Is there a missing feature you want to implement? Please briefly describe it and
  the problem it solves.

- Did you find some other issue with deal.II? Summarize the problem here.
-->

Interesting question - we should document these restrictions better - some of these functions, as far as I can tell, don't make sense with nonzero codimension. In particular, if I recall correctly, this constructor was added to support some other matrix-free code (which does not support nonzero codimension problems yet).

If I try to compile
```cpp
      constexpr int order    = 2;
      constexpr int dim      = 2;
      constexpr int spacedim = 3;

      Tensor<order, spacedim, Tensor<1, dim>> t;

      auto d2 = DerivativeForm<order, dim, spacedim>(t);
```

I get the correct compilation error
```
/home/drwells/Documents/Code/CPP/dealii-dev/include/deal.II/base/derivative_form.h: In instantiation of 
    'dealii::DerivativeForm<order, dim, spacedim, Number>::DerivativeForm(
        const dealii::Tensor<order, spacedim, dealii::Tensor<1, dim, Number> >&) 
    [with int order = 2; int dim = 2; int spacedim = 3; Number = double]':
/home/drwells/Documents/Code/CPP/dealii-experiments/derivative-form-bug/bug.cc:34:55:   required from here
/home/drwells/Documents/Code/CPP/dealii-dev/include/deal.II/base/derivative_form.h:205:16: error: 
    no match for 'operator=' (operand types are 'dealii::Tensor<2, 2, double>' and 
    'const dealii::Tensor<2, 3, dealii::Tensor<1, 2, double> >::value_type' 
    {aka 'const dealii::Tensor<1, 3, dealii::Tensor<1, 2, double> >'})
  205 |     (*this)[j] = T[j];
      |     ~~~~~~~~~~~^~~~
```

Since it's a template class, compilation won't fail until one tries to call that constructor with the 'wrong' types. Were you able to get farther than this? I believe this should fail to compile in all nonzero codimension cases.

--- a/doc/news/changes/minor/20230201Bangerth
+++ b/doc/news/changes/minor/20230201Bangerth
@@ -1,0 +1,7 @@
+Fixed: The DerivativeForm class had a constructor that took an array
+of elements via the type
+`Tensor<order, spacedim, Tensor<1, dim,Number>>`.
+This should have been
+`Tensor<1, spacedim, Tensor<order, dim, Number>>`, and it is now fixed.
+<br>
+(Robin Goermer, Wolfgang Bangerth, 2023/02/01)

--- a/include/deal.II/base/derivative_form.h
+++ b/include/deal.II/base/derivative_form.h
@@ -69,9 +69,12 @@ public:
   DerivativeForm(const Tensor<order + 1, dim, Number> &);
 
   /**
-   * Constructor from a tensor.
+   * Constructor. This constructor initializes the data stored by this
+   * object from a `spacedim x dim^order` array that is represented
+   * by a Tensor with `spacedim` components each of which is a
+   * Tensor of rank `order` and size `dim`.
    */
-  DerivativeForm(const Tensor<order, spacedim, Tensor<1, dim, Number>> &);
+  DerivativeForm(const Tensor<1, spacedim, Tensor<order, dim, Number>> &);
 
   /**
    * Read-Write access operator.
@@ -199,7 +202,7 @@ inline DerivativeForm<order, dim, spacedim, Number>::DerivativeForm(
 
 template <int order, int dim, int spacedim, typename Number>
 inline DerivativeForm<order, dim, spacedim, Number>::DerivativeForm(
-  const Tensor<order, spacedim, Tensor<1, dim, Number>> &T)
+  const Tensor<1, spacedim, Tensor<order, dim, Number>> &T)
 {
   for (unsigned int j = 0; j < spacedim; ++j)
     (*this)[j] = T[j];

--- a/tests/base/derivative_form_initialization.cc
+++ b/tests/base/derivative_form_initialization.cc
@@ -1,0 +1,60 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Verify that DerivativeForm can be initialized from a
+// Tensor<1,dim,Tensor<order,spacedim>. We used to have a bug where the
+// copy constructor offered Tensor<order,dim,Tensor<1,spacedim> instead,
+// but that was just wrong.
+
+#include <deal.II/base/derivative_form.h>
+
+#include <boost/type_traits.hpp>
+
+#include <complex>
+
+#include "../tests.h"
+
+template <typename Number>
+void
+test()
+{
+  // Choose orders and dimensions so that they are all not equal:
+  constexpr int order    = 2;
+  constexpr int dim      = 1;
+  constexpr int spacedim = 3;
+
+
+  Tensor<order, dim, Number> inner;
+  inner[0][0] = 1;
+  Tensor<1, spacedim, Tensor<order, dim, Number>> t;
+  t[0] = t[1] = t[2] = inner;
+
+  const DerivativeForm<order, dim, spacedim, Number> d(t);
+
+  deallog << "norm: " << d.norm() << std::endl;
+}
+
+int
+main()
+{
+  initlog();
+
+  deallog << "testing float" << std::endl;
+  test<float>();
+
+  deallog << "testing double" << std::endl;
+  test<double>();
+}

--- a/tests/base/derivative_form_initialization.output
+++ b/tests/base/derivative_form_initialization.output
@@ -1,0 +1,5 @@
+
+DEAL::testing float
+DEAL::norm: 1.73205
+DEAL::testing double
+DEAL::norm: 1.73205

